lib/ABI/FunctionType/Layout.cpp

```diff
@@ -0,0 +1,963 @@
+
+          if (MaybeFieldSize.value() != MaybeRegisterSize.value())
+            break;
+
+          auto Tie = std::tie(Register, Field.Type());
+          auto [Iterator, Success] = RecoveredTypes.insert(std::move(Tie));
+          revng_assert(Success);
+
+          StructOffset += MaybeFieldSize.value();
+        }
+
+        // Only preserve types if there's one for each of the registers.
+        // The worst thing we could do is to inject incompatible information
+        // in.
+        if (RecoveredTypes.size() == NewType.ReturnValues().size()) {
+          for (auto [Register, Type] : RecoveredTypes)
+            NewType.ReturnValues().at(Register).Type() = Type;
+
+          revng_log(Log, "The preservation attempt was successful.");
+        } else {
+          revng_log(Log, "The preservation attempt was not successful.");
+        }
+      }
+    }
+  } else if (ReturnValue.Size != 0) {
+    // The return value uses a pointer-to-a-location: add it as an argument.
+
+    auto MaybeReturnValueSize = Function.ReturnType().size();
+    revng_assert(MaybeReturnValueSize != std::nullopt);
+    revng_assert(ReturnValue.Size == *MaybeReturnValueSize);
+
+    model::QualifiedType ReturnType = Function.ReturnType();
+    ReturnType.Qualifiers().emplace_back(PointerQualifier);
+
+    revng_assert(!ABI.GeneralPurposeReturnValueRegisters().empty());
+    auto FirstRegister = ABI.GeneralPurposeReturnValueRegisters()[0];
+    model::TypedRegister ReturnPointer(FirstRegister);
+    ReturnPointer.Type() = std::move(ReturnType);
+    NewType.ReturnValues().insert(std::move(ReturnPointer));
+  } else {
+    // The function returns `void`: no need to do anything special.
+  }
+
+  model::StructType StackArguments;
+  uint64_t CurrentStackOffset = 0;
+
+  // Now that return value is figured out, the arguments are next.
+  auto Arguments = distributeArguments(Function.Arguments(),
+                                       ReturnValue.SizeOnStack != 0);
+  std::size_t Index = 0;
+  for (const DistributedArgument &Distributed : Arguments) {
+    if (!Distributed.RepresentsPadding) {
+      // Transfer the register arguments first.
+      for (model::Register::Values Register : Distributed.Registers) {
+        model::NamedTypedRegister Argument(Register);
+
+        auto M = addNewTypes(*Binary);
+        bool SimpleArgument = Distributed.Registers.size() == 1
+                              && Distributed.SizeOnStack == 0;
+        if (SimpleArgument) {
+          const auto &ArgumentType = Function.Arguments().at(Index).Type();
+          revng_assert(*ArgumentType.size() <= ABI.getPointerSize());
+          Argument.Type() = chooseType(ArgumentType, Register, M);
+
+          // TODO: see what can be done to preserve names better
+          auto ArgumentName = Function.Arguments().at(Index).name();
+          if (llvm::StringRef{ ArgumentName.str() }.take_front(8) != "unnamed_")
+            Argument.CustomName() = ArgumentName;
```

Comment:
Closing the thread in favour of [this one](https://github.com/revng/revng/pull/232#discussion_r1101154834) since it's a more broad view on the same situation.